### ISSUE TITLE
Use common investment rule function

### ISSUE THREAD

**Issue by nailend** [oemof/oemof-solph#958]
There are duplicated rule formulations in the investment blocks of  [`flows`](https://github.com/oemof/oemof-solph/blob/1cee273d3328fe00daf092f6bdfe9d227573a443/src/oemof/solph/components/_generic_storage.py#L1220), [`_generic_storage`](https://github.com/oemof/oemof-solph/blob/1cee273d3328fe00daf092f6bdfe9d227573a443/src/oemof/solph/components/_generic_storage.py#L1220) and [`SinkDSM` which](https://github.com/oemof/oemof-solph/blob/1cee273d3328fe00daf092f6bdfe9d227573a443/src/oemof/solph/components/experimental/_sink_dsm.py#L964C47-L964C47) only differ slightly with their indexes and constraint name. This counts for:

* _old_storage_capacity_rule_end
* _old_dsm_capacity_rule_end (for every SinkDSM approach)
* _old_capacity_rule_end

Is there a way to define a common function which is called inside and `indexes` and `constraint` are passed. The advantage would be, that if anything changes, not five different code-locations have to be changed and checked.

**Comment by p-snft** [comment]
Thanks for bringing this up. I'd suggest to split _options.py into a submodule. The constraints would then be formulated in options/_investment.py that bundles the class `Investment` and the constraints (and analogue for `NonConvex`).

**Comment by jokochems** [comment]
> There are duplicated rule formulations in the investment blocks of [`flows`](https://github.com/oemof/oemof-solph/blob/1cee273d3328fe00daf092f6bdfe9d227573a443/src/oemof/solph/components/_generic_storage.py#L1220), [`_generic_storage`](https://github.com/oemof/oemof-solph/blob/1cee273d3328fe00daf092f6bdfe9d227573a443/src/oemof/solph/components/_generic_storage.py#L1220) and [`SinkDSM`](https://github.com/oemof/oemof-solph/blob/1cee273d3328fe00daf092f6bdfe9d227573a443/src/oemof/solph/components/experimental/_sink_dsm.py#L964C47-L964C47) which only differ slightly with their indexes and constraint name. This counts for:
> 
> * _old_storage_capacity_rule_end
> * _old_dsm_capacity_rule_end (for every SinkDSM approach)
> * _old_capacity_rule_end
> 
> Is there a way to define a common function which is called inside and `indexes` and `constraint` are passed. The advantage would be, that if anything changes, not five different code-locations have to be changed and checked. I just encountered this problem in #957

I agree that this would be sensible and facilitate code maintenance as well as development of new functionality.
